### PR TITLE
[cu-2f7bbjk] Creating LP Staking Config From Pablo

### PR DESCRIPTION
**Pull request by vimukthi-git**
## Issue
https://app.clickup.com/t/2f7bbjk

## Description
Allow LPs to stake their LP tokens to earn PICA and PBLO

## Checklist

- [x] I have updated the cargo docs to reflect changes made by this PR _(if applicable)_
- [x] I have updated the `book/` to reflect changes made by this PR _(if applicable)_

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2f7bbjk 5.2.2.1 Expose interface for LP fNFT Configuration and Configure fNFT at Pool Creation](https://app.clickup.com/t/2f7bbjk)

File: frame/pablo/src/lib.rs [hunk 1]
```diff
@@ -86,21 +86,35 @@ pub mod pallet {
...
 	};
+	use primitives::currency::CurrencyId;
+	use sp_arithmetic::Perbill;
 	use sp_runtime::{
 		traits::{AccountIdConversion, BlockNumberProvider, Convert, One, Zero},
 		ArithmeticError, FixedPointNumber, Permill,
 	};
 	use sp_std::vec::Vec;
 
+	pub const MINUTE: DurationSeconds = 60;
```
Comment (dzmitry-lahoda) [inline frame/pablo/src/lib.rs:112]: please move it into time.rs

File: frame/pablo/src/liquidity_bootstrapping.rs [hunk 2]
```diff
@@ -76,7 +76,7 @@ impl<T: Config> LiquidityBootstrapping<T> {
 				Ok(pool_id)
 			})?;
 
-		Ok(pool_id)
+		Ok((pool_id, T::AssetId::default()))
```
Comment (dzmitry-lahoda) [inline frame/pablo/src/liquidity_bootstrapping.rs:79]: not sure, can you leave TODO here to make use of currency-factory?

File: frame/pablo/src/lib.rs [hunk 3]
```diff
@@ -254,7 +268,8 @@ pub mod pallet {
 			+ Debug
 			+ Default
 			+ TypeInfo
-			+ Ord;
+			+ Ord
+			+ From<CurrencyId>;
```
Comment (dzmitry-lahoda) [inline frame/pablo/src/lib.rs:272]: I think where headed to wards interfaces

like 
trait SpecialAssets {
    fn get_local_native() -> 
    fn get_relay_native() ->
    fn get_protocol_native(pallet_id) -> Option<AssetId>
}

File: frame/pablo/src/lib.rs [hunk 4]
```diff
@@ -586,6 +607,18 @@ pub mod pallet {
 					)
 				},
 			};
+			T::StakingConfiguration::configure(
+				lp_token,
+				// TODO (vim) Hard coded for now
+				[(WEEK, Perbill::from_float(0.5)), (MONTH, Perbill::from_float(1.0))].into(),
+				// TODO (vim) : Consider the case of having change the rewarded assets after the
+				// NFTs have already been created for a given asset by stakers
+				BTreeSet::from([CurrencyId::PBLO.into(), CurrencyId::PICA.into()]),
```
Comment (dzmitry-lahoda) [inline frame/pablo/src/lib.rs:616]: see my comment about trait based currency

File: runtime/primitives/src/currency.rs [hunk 5]
```diff
@@ -76,6 +76,9 @@ impl CurrencyId {
 	#[allow(non_upper_case_globals)]
 	pub const kUSD: CurrencyId = CurrencyId(129);
 
+	/// Protocol Tokens
+	pub const PBLO: CurrencyId = CurrencyId(201);
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply runtime/primitives/src/currency.rs:80]: yes. good enough. for static protocols static factory is good enough.

File: frame/pablo/src/lib.rs (hunk 3 above)
Comment (vimukthi-git) [reply frame/pablo/src/lib.rs:272]: Good point, Though may be we should wait for a few more use cases before doing such refactorings? Specially the use here is to set rewarded asset types, which can be much more than just what you list out here.

File: frame/pablo/src/liquidity_bootstrapping.rs (hunk 2 above)
Comment (vimukthi-git) [reply frame/pablo/src/liquidity_bootstrapping.rs:79]: This is just a placeholder value to for LBP, which does not really have an LP token. But you are right in that I need to skip this value when creating the staking config

**Comment by vimukthi-git** [comment]
reviewers note: This PR is currently blocked until I find a way to avoid the hard dependency on staking rewards for now in dali runtime for pablo.

File: frame/staking-rewards/src/lib.rs [hunk 6]
```diff
@@ -611,6 +614,28 @@ pub mod pallet {
...
+	impl<T: Config> StakingConfiguration for Pallet<T> {
+		type AssetId = AssetIdOf<T>;
+		type AccountId = AccountIdOf<T>;
+
+		fn configure(
+			asset: Self::AssetId,
+			duration_presets: BTreeMap<DurationSeconds, Perbill>,
+			reward_assets: BTreeSet<Self::AssetId>,
+			early_unstake_penalty: Penalty<Self::AccountId>,
+		) -> Result<(), DispatchError> {
```
Comment (dzmitry-lahoda) [inline frame/staking-rewards/src/lib.rs:626]: may manage having some config and consistent updates to it

Comment (dzmitry-lahoda) [inline frame/staking-rewards/src/lib.rs:626]: may manage having some config and consistent updates to it

Comment (vimukthi-git) [reply frame/staking-rewards/src/lib.rs:626]: Could you clarify what you mean? Is there any update to make here?

**Comment by vimukthi-git** [comment]
Closing this PR as it's blocked by updates to new staking rewards pallet. Will reopen later when it is possible to make progress.

**Pull request closed without merging**